internal/provider/data_source_machine.go

```diff
@@ -2,59 +2,139 @@ package provider
 				Required:    true,
 			},
-			"machine_id": {
+			"machine_id": schema.StringAttribute{
 				Description: "The Juju id of the machine.",
-				Type:        schema.TypeString,
 				Required:    true,
 			},
+			// ID required by the testing framework
+			"id": schema.StringAttribute{
+				Computed: true,
+			},
 		},
 	}
 }
 
-func dataSourceMachineRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+// Configure enables provider-level data or clients to be set in the
+// provider-defined DataSource type. It is separately executed for each
+// ReadDataSource RPC.
+func (d *machineDataSource) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
 
-	modelName := d.Get("model").(string)
-	machine_id := d.Get("machine_id").(string)
+	client, ok := req.ProviderData.(*juju.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *juju.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
 
-	model, err := client.Models.GetModelByName(modelName)
-	if err != nil {
-		return diag.FromErr(err)
+	d.client = client
+	d.subCtx = tflog.NewSubsystem(ctx, LogDataSourceMachine)
+}
+
+// Read is called when the provider must read data source values in
+// order to update state. Config values should be read from the
+// ReadRequest and new state values set on the ReadResponse.
+func (d *machineDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	// Prevent panic if the provider has not been configured.
+	if d.client == nil {
+		addDSClientNotConfiguredError(&resp.Diagnostics, "machine")
+		return
 	}
 
-	machine, err := client.Machines.ReadMachine(&juju.ReadMachineInput{
-		ModelUUID: model.UUID,
-		MachineId: machine_id,
-	})
+	var data machineDataSourceModel
+
+	// Read Terraform configuration data into the model.
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
+	// Get current juju machine data source values.
+	// "id" matches previous provider values however is not
+	// unique and is only used for tests. Data sources cannot
+	// be imported by terraform.
+	model, err := d.client.Models.GetModelByName(data.Model.ValueString())
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read model, got error: %s", err))
+		return
 	}
+	machine_id := data.MachineID.ValueString()
+	d.trace(fmt.Sprintf("reading juju machine %q data source", machine_id))
 
-	d.SetId(machine.MachineId)
-	if err = d.Set("model", model.Name); err != nil {
-		return diag.FromErr(err)
+	// Verify the machine exists in the model provided
+	if _, err = d.client.Machines.ReadMachine(
```

Comment:
I thought it was better to use ReadMachine rather than FullStatus, as the results can be very large and go thru the model cache depending on juju version, thus can be behind in data.